From d4c464cfe1a9840dd156107ef2a4ae553a92d922 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:36:42 -0400
Subject: [block] blk-mq: always initialize request->start_time

Message-id: <1402670238-13916-115-git-send-email-snitzer@redhat.com>
Patchwork-id: 83891
O-Subject: [RHEL7.1 PATCH 114/150] blk-mq: always initialize request->start_time
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit 3ee3237239583a6555db4f297d00eebdbb6d76ad
Author: Jens Axboe <axboe@fb.com>
Date:   Mon Jun 9 09:36:53 2014 -0600

    blk-mq: always initialize request->start_time

    The blk-mq core only initializes this if io stats are enabled, since
    blk-mq only reads the field in that case. But drivers could
    potentially use it internally, so ensure that we always set it to
    the current time when the request is allocated.

    Reported-by: Ming Lei <tom.leiming@gmail.com>
    Signed-off-by: Jens Axboe <axboe@fb.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-mq.c b/block/blk-mq.c
index fafea52..a5ea37d 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -185,6 +185,7 @@ static void blk_mq_rq_ctx_init(struct request_queue *q, struct blk_mq_ctx *ctx,
  RB_CLEAR_NODE(&rq->rb_node);
  rq->rq_disk = NULL;
  rq->part = NULL;
+ rq->start_time = jiffies;
 #ifdef CONFIG_BLK_CGROUP
  rq->rl = NULL;
  set_start_time_ns(rq);
@@ -1104,10 +1105,8 @@ static void blk_mq_bio_to_request(struct request *rq, struct bio *bio)
 {
  init_request_from_bio(rq, bio);
 
- if (blk_do_io_stat(rq)) {
-  rq->start_time = jiffies;
+ if (blk_do_io_stat(rq))
   blk_account_io_start(rq, 1);
- }
 }
 
 static inline bool blk_mq_merge_queue_io(struct blk_mq_hw_ctx *hctx,
-- 
1.7.1